### PR TITLE
rm: linux: add deeper information about lmp updates

### PR DESCRIPTION
**Pull request by vanmaegima**
## Readiness

* [ ] Merge (pending reviews)
* [ ] Merge after _date or event_
* [ ] Draft

## Overview

_Why merge this PR? What does it solve?_

## Checklist

_Optional. Add a 'x' to steps taken._
_You can fill this out after opening the PR. "Did I..."_

* [ ] Run spelling and grammar check, preferably with linter.
* [ ] Avoid changing any header associated with a link/reference.
* [ ] Step through instructions (or ask someone to do so).
* [ ] Review for [wordiness](https://languagetool.org/insights/post/wordiness/)
* [ ] Match tone and style of page/section.
* [ ] Run `make linkcheck`.
* [ ] View HTML in a browser to check rendering.
* [ ] Use [semantic newlines](https://bobheadxi.dev/semantic-line-breaks/).
* [ ] follow best practices for commits.
  * [ ] Descriptive title written in the imperative.
  * [ ] Include brief overview of QA steps taken.
  * [ ] Mention any related issues numbers.
  * [ ] End message with sign off/DCO line (`-s, --signoff`).
  * [ ] Sign commit with your gpg key (`-S, --gpg-sign`).
  * [ ] Squash commits if needed.
* [ ] Request PR review by a technical writer and at least one peer.

## Comments

_Any thing else that a maintainer/reviewer should know._
_This could include potential issues, rational for approach, etc._

### CODE REVIEW

**Comment by doanac** [comment]
Docs for 58810596df80dff1331af2533bbbc357f4692b84 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2489/docs/artifacts/html/index.html

File: source/reference-manual/linux/linux-update.rst [hunk 1]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+    $ git clone https://github.com/foundriesio/lmp-tools
+    $ cd lmp-manifest/
+    $ git checkout <branch to update>
+    $ ../lmp-tools/scripts/update-factory-manifest
+    New upstream release(s) have been found.
+    Merging local code with upstream release: 91
+    Proceed ? (y/n): y
+
+If no merge conflicts are found, it merges your changes and pushes the updated manifest to your Factory, triggering a new platform build.
+Once published, the update is deployed to your devices following this tag.
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:52]:
```suggestion
Once published, the update is deployed to devices following this tag.
```

File: source/reference-manual/linux/linux-update.rst [hunk 2]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+In this case, you need to identify the conflicting commit and revert this change before proceeding:
+
+.. code-block::
+
+    $ git revert f54f9f9939b6c1548f838e9bd85b98bb01a6a42f
+    $ ../lmp-tools/scripts/update-factory-manifest
+    Automatic update successful!
+
+.. tip::
+    If something goes wrong, don’t fret! This is why we use version control!
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:81]:
```suggestion
    If something goes wrong, do not fret! This is why we use version control!
```

File: source/reference-manual/linux/linux-update.rst [hunk 3]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+
+    .. code-block::
+
+        $ git revert HEAD
+
+
+``meta-subscriber-overrides``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The previous step triggers a new platform build for the latest LmP release in your FoundriesFactory.
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:91]:
```suggestion
The previous step triggers a new platform build for the latest LmP release in your Factory.
```

File: source/reference-manual/linux/linux-update.rst [hunk 4]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+
+.. code-block::
+
+    $ git clone git://github.com/foundriesio/u-boot.git # SRC_URI
+    $ cd u-boot
+    $ git checkout d5976b6253dcae875fb42fbef68e1d05e7de5141 # SRCREV
+
+a. Patches
+
+In some cases, customer Factories carry U-Boot code that has been applied in LmP sources.
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:171]:
```suggestion
In some cases, Factories carry U-Boot code that has been applied in LmP sources.
```

File: source/reference-manual/linux/linux-update.rst [hunk 5]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+.. important::
+    Remember to trigger :ref:`ref-boot-software-updates` when necessary.
+
+1. Take a bench device and flash it with the latest stable image of the **old** LmP version (e.g. v88).
+
+2. Register it to the Factory to the tag which brings the new LmP version, for example ``next`` (e.g. v91):
+
+.. code-block::
+
+    $ lmp-device-register -n test-lmp-update -t next
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:382]:
```suggestion
    .. code-block::

        $ lmp-device-register -n test-lmp-update -t next
```

File: source/reference-manual/linux/linux-update.rst [hunk 6]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+Once your ``next`` branch is in a good state, you may wish to migrate your development branches to this new release.
+Here, the development branch is called ``devel``.
+
+1. Clone all 3 required repos:
+
+.. code-block::
+
+    $ git clone https://github.com/foundriesio/lmp-tools
+    $ git clone https://source.foundries.io/factories/<YOUR FACTORY>/lmp-manifest
+    $ git clone https://source.foundries.io/factories/<YOUR FACTORY>/meta-subscriber-overrides
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:400]:
```suggestion
    .. code-block::

        $ git clone https://github.com/foundriesio/lmp-tools
        $ git clone https://source.foundries.io/factories/<YOUR FACTORY>/lmp-manifest
        $ git clone https://source.foundries.io/factories/<YOUR FACTORY>/meta-subscriber-overrides
```

File: source/reference-manual/linux/linux-update.rst [hunk 7]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+    $ git checkout <branch to update>
+    $ ../lmp-tools/scripts/update-factory-manifest
+    New upstream release(s) have been found.
+    Merging local code with upstream release: 91
+    Proceed ? (y/n): y
+
+If no merge conflicts are found, it merges your changes and pushes the updated manifest to your Factory, triggering a new platform build.
+Once published, the update is deployed to your devices following this tag.
+
+Common Problems
```
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:54]: I would say "Common Pitfalls" instead.

File: source/reference-manual/linux/linux-update.rst [hunk 8]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+.. code-block::
+
+    $ cd meta-subscriber-overrides
+    $ git checkout next
+    $ git pull --rebase
+    $ git checkout devel
+    $ git pull --rebase
+    $ git merge --ff-only next
+    $ git commit --allow-empty -m "[skip ci] Update to LmP v91"
+    $ git push
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:413]:
```suggestion
    .. code-block::

        $ cd meta-subscriber-overrides
        $ git checkout next
        $ git pull --rebase
        $ git checkout devel
        $ git pull --rebase
        $ git merge --ff-only next
        $ git commit --allow-empty -m "[skip ci] Update to LmP v91"
        $ git push
```

File: source/reference-manual/linux/linux-update.rst [hunk 9]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+    Merging local code with upstream release: 91
+    Proceed ? (y/n): y
+
+If no merge conflicts are found, it merges your changes and pushes the updated manifest to your Factory, triggering a new platform build.
+Once published, the update is deployed to your devices following this tag.
+
+Common Problems
+"""""""""""""""
+
+The most common problems found at this stage are related to manual changes to ``lmp-manifest``.
```
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:57]:
```suggestion
Usually at this stage, the problems are related to manual changes to ``lmp-manifest``.
```

However, I would drop "manual" and say only "changes". As we cannot control how the changes are made

File: source/reference-manual/linux/linux-update.rst [hunk 10]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+
+.. code-block::
+
+    $ cd lmp-manifest
+    $ git checkout devel
+    $ git pull --rebase
+    $ ../lmp-tools/scripts/update-factory-manifest
+    New upstream release(s) have been found.
+    Merging local code with upstream release: 91
+    Proceed ? (y/n):
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:425]:
```suggestion
    .. code-block::

        $ cd lmp-manifest
        $ git checkout devel
        $ git pull --rebase
        $ ../lmp-tools/scripts/update-factory-manifest
        New upstream release(s) have been found.
        Merging local code with upstream release: 91
        Proceed ? (y/n):
```

File: source/reference-manual/linux/linux-update.rst [hunk 11]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+
+4. Proceed by typing ``y``. This updates the ``lmp-manifest/devel`` branch and trigger a build for the new release.
+
+5. Once it is built, a new Target for the latest LmP release becomes available for your development devices following ``devel``.
+
+Common Errors and Tips
+~~~~~~~~~~~~~~~~~~~~~~
+
+* A good practice when debugging migration issues is to compare the reference machine changes from one LmP version to the other. Likely, the changes from the reference machine should be mirrored to your custom machine.
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:434]:
```suggestion
* A good practice when debugging migration issues is to compare the reference machine changes from one LmP version to the other.
  Likely, the changes from the reference machine should be mirrored to your custom machine.
```

File: source/reference-manual/linux/linux-update.rst [hunk 12]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+    Proceed ? (y/n): y
+
+If no merge conflicts are found, it merges your changes and pushes the updated manifest to your Factory, triggering a new platform build.
+Once published, the update is deployed to your devices following this tag.
+
+Common Problems
+"""""""""""""""
+
+The most common problems found at this stage are related to manual changes to ``lmp-manifest``.
+For example, if you have manually bumped ``meta-lmp`` in order to bring in new changes, it causes a merge conflict:
```
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:58]:
```suggestion
For example, if you have bumped only ``meta-lmp`` to bring in new changes, it causes a merge conflict:
```

File: source/reference-manual/linux/linux-update.rst [hunk 13]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+Common Errors and Tips
+~~~~~~~~~~~~~~~~~~~~~~
+
+* A good practice when debugging migration issues is to compare the reference machine changes from one LmP version to the other. Likely, the changes from the reference machine should be mirrored to your custom machine.
+
+* Working on the LmP update in a separate branch is highly recommended so it does not block your development branches.
+
+* For machines that support :ref:`lmp-mfgtool distro <ref-lmp-mfgtool>`, use that for a quick debug iteration: there is no need to flash the whole image to verify U-Boot, for example.
+
+* Also for machines that support :ref:`lmp-mfgtool distro <ref-lmp-mfgtool>`, the suggestion is to keep a single source of patches for hardware support (for ``u-boot-fio``/``u-boot-fio-mfgtool`` and ``linux-lmp-fslc-imx``/``linux-lmp-dev-mfgtool``). This avoids duplicated code in the Factory. Example:
```
Comment (kprosise) [inline source/reference-manual/linux/linux-update.rst:440]:
```suggestion
* Also for machines that support :ref:`lmp-mfgtool distro <ref-lmp-mfgtool>`, the suggestion is to keep a single source of patches for hardware support (for ``u-boot-fio``/``u-boot-fio-mfgtool`` and ``linux-lmp-fslc-imx``/``linux-lmp-dev-mfgtool``). This avoids duplicated code in the Factory.
For example:
```

File: source/reference-manual/linux/linux-update.rst (hunk 3 above)
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:91]:
```suggestion
The previous step pushes the update to the FoundriesFactory, which triggers a new platform build for the latest LmP release.
```

**Review by kprosise — COMMENTED** [review]
First pass, I will do a follow up review before we merge.

File: source/reference-manual/linux/linux-update.rst [hunk 14]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+    .. code-block::
+
+        $ git revert HEAD
+
+
+``meta-subscriber-overrides``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The previous step triggers a new platform build for the latest LmP release in your FoundriesFactory.
+During the migration, this initial build often fails as ``meta-subcriber-overrides`` may have changes that need to be adjusted to the new release.
```
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:92]:
```suggestion
It is expected that, during the migration, this initial build fails as ``meta-subcriber-overrides`` may have changes that require adjustments to the new release.
```

File: source/reference-manual/linux/linux-update.rst [hunk 15]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+
+    $ cat meta-lmp/meta-lmp-bsp/conf/layer.conf | grep COMPAT
+    LAYERSERIES_COMPAT_meta-lmp-bsp = "kirkstone"
+    $ cat meta-subscriber-overrides/conf/layer.conf | grep COMPAT
+    LAYERSERIES_COMPAT_meta-subscriber-overrides = "kirkstone"
+
+* **Revert changes**
+
+In some cases, we need to bring LmP post-release changes into the Factory to fix critical bugs or to bring in a new feature.
+An example is to bump projects.
```
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:113]: to bump projects or to bump packages?

File: source/reference-manual/linux/linux-update.rst [hunk 16]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+
+    diff --git a/recipes-sota/lmp-device-register/lmp-device-register_git.bbappend b/recipes-sota/lmp-device-register/lmp-device-register_git.bbappend
+    new file mode 100644
+    index 0000000..0bdbd23
+    --- /dev/null
+    +++ b/recipes-sota/lmp-device-register/lmp-device-register_git.bbappend
+    @@ -0,0 +1 @@
+    +SRCREV = "848bcbbba886320b13b11ac04826be0361288619"
+
+The recommendation is to tag commits like this with a ``[REVERTME-vXX]`` tag that indicates which LmP version they should be reverted.
```
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:130]: I don't understand this sentence.
Why am I reverting a bump? I don't get why a bump is problematic? (I think I missed something in this explanation)

File: source/reference-manual/linux/linux-update.rst [hunk 17]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+.. code-block::
+
+    $ git clone git://github.com/foundriesio/u-boot.git # SRC_URI
+    $ cd u-boot
+    $ git checkout d5976b6253dcae875fb42fbef68e1d05e7de5141 # SRCREV
+
+a. Patches
+
+In some cases, customer Factories carry U-Boot code that has been applied in LmP sources.
+These patches do not apply cleanly to the sources and cause build errors (`do_patch` fails).
```
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:172]:
```suggestion
Customer Factories sometimes carry U-Boot code that has been applied in LmP sources.
These patches do not apply cleanly to the sources and cause build errors (`do_patch` fails).
```

File: source/reference-manual/linux/linux-update.rst [hunk 18]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+
+    # v88
+    $ find ../layers/ -iname linux-lmp-dev-mfgtool*
+    ../layers/meta-lmp/meta-lmp-base/recipes-kernel/linux/linux-lmp-dev-mfgtool.bb
+    ../layers/meta-lmp/meta-lmp-bsp/recipes-kernel/linux/linux-lmp-dev-mfgtool.bbappend
+    ../layers/meta-lmp/meta-lmp-bsp/recipes-kernel/linux/linux-lmp-dev-mfgtool
+
+    # v91
+    $ find ../layers/ -iname linux-lmp-dev-mfgtool*
+    ../layers/meta-lmp/meta-lmp-base/recipes-kernel/linux/linux-lmp-dev-mfgtool_git.bb
```
Comment (angolini) [inline source/reference-manual/linux/linux-update.rst:509]: Now I get what you mean. All recipes are versioned! The difference is that now the version is explicitly set on file name.

**Review by angolini — COMMENTED** [review]
Great content! Great text! Only few minor suggestions

**Review by angolini — APPROVED** [review]

File: source/reference-manual/linux/linux-update.rst (hunk 11 above)
Comment (kprosise) [reply source/reference-manual/linux/linux-update.rst:434]: Apologies, GitHub formatting is a pain, I'll make any format changes locally later on, as a follow up after this is merged.

**Comment by doanac** [comment]
Docs for 667d39cdde22c024f9d7e50ac933b10beae88493 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2490/docs/artifacts/html/index.html

File: source/reference-manual/linux/linux-update.rst (hunk 15 above)
Comment (vanmaegima) [reply source/reference-manual/linux/linux-update.rst:113]: @angolini I'm not sure what's the correct wording here. We usually do aktualizr bumps for example.

File: source/reference-manual/linux/linux-update.rst (hunk 16 above)
Comment (vanmaegima) [reply source/reference-manual/linux/linux-update.rst:130]: @angolini If we do a bump with a new SRCREV in meta-subscriber-overrides, then what's set in meta-lmp will be overwritten, that's what I meant
Any suggestions to make this clearer? This is something normal most customers do and we revert once changes are applied to their current lmp versions
Just a note: we will probably have another doc about applying patches/bumping recipes from lmp due to the new source code policies, after that we can link a page here

File: source/reference-manual/linux/linux-update.rst [hunk 19]
```diff
@@ -12,30 +12,507 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+    SRC_URI="git://github.com/Freescale/linux-fslc.git;protocol=https;branch=5.10-2.1.x-imx;name=machine; \
+    git://github.com/foundriesio/lmp-kernel-cache.git;protocol=https;type=kmeta;name=meta;branch=linux-v5.10.y;destsuffix=kernel-meta \
+    file://0004-FIO-toup-hwrng-optee-support-generic-crypto.patch \
+    file://0001-FIO-extras-arm64-dts-imx8mm-evk-use-imx8mm-evkb-for-.patch \
+    file://0001-FIO-tonxp-drm-bridge-it6161-add-missing-gpio-consume.patch \
+    file://0001-arm64-dts-imx8mq-drop-cpu-idle-states.patch \
+    file://0001-FIO-temphack-ARM-mach-imx-conditionally-disable-some.patch \
+    file://0001-FIO-internal-arm64-dts-imx8mn-evk.dtsi-re-add-blueto.patch "
+
+* Sometimes, a core recipe can change names or become versioned. In this case, old `.bbappends` may fail to override the recipe, for example:
```
[the remark this answers is not in the document]
Comment (vanmaegima) [reply source/reference-manual/linux/linux-update.rst:491]: maybe I can say a recipe was renamed?

**Comment by vanmaegima** [comment]
@kprosise @angolini Thank you for the quick review! I should have addressed all open comments, just waiting on open discussions with Daiane.
I will wait for one review from the CS team before we merge this (having this in v92 would be great though).

File: source/reference-manual/linux/linux-update.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (angolini) [reply source/reference-manual/linux/linux-update.rst:491]: I prefer to way only "recipe was renamed"

**Comment by doanac** [comment]
Docs for 991717878f176f944b609b349536a90c634ae75b are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2494/docs/artifacts/html/index.html

File: source/reference-manual/linux/linux-update.rst [hunk 20]
```diff
@@ -12,30 +12,515 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+    $ cd lmp-manifest/
+    $ git checkout <branch to update>
+    $ ../lmp-tools/scripts/update-factory-manifest
+    New upstream release(s) have been found.
+    Merging local code with upstream release: 91
+    Proceed ? (y/n): y
+
+If no merge conflicts are found, it merges your changes and pushes the updated manifest to your Factory, triggering a new platform build.
+
+.. warning::
```
Comment (caiotpereira) [inline source/reference-manual/linux/linux-update.rst:53]: Good Remark!

File: source/reference-manual/linux/linux-update.rst [hunk 21]
```diff
@@ -12,30 +12,515 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+
+    $ cd meta-lmp
+    $ git diff mp-88 mp-91 meta-lmp-bsp/recipes-security/optee/optee-os-fio-bsp.inc
+
+Bring relevant changes from the reference machine to your machine code.
+
+* **Mfgtool** (if applicable)
+
+.. note::
+    Not all machines require/support ``mfgtool`` build. Currently, i.MX and STM32MP are supported.
```
Comment (caiotpereira) [inline source/reference-manual/linux/linux-update.rst:339]: I wonder if the TI mfg tool was officially released.

Comment (vanmaegima) [reply source/reference-manual/linux/linux-update.rst:339]: yes, it is part of lmp

File: source/reference-manual/linux/linux-update.rst [hunk 22]
```diff
@@ -12,30 +12,515 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
 Updating Your Factory
 ~~~~~~~~~~~~~~~~~~~~~
 
-.. prompt:: bash
+.. tip::
+    We suggest that you perform the release migration on a new branch, forked off your stable development, so as to avoid impacting your current devices.
+    After testing, the changes can be merged back to development.
+
+    See :ref:`ref-new-branch`.
+
```
[the remark this answers is not in the document]
Comment (vanmaegima) [reply source/reference-manual/linux/linux-update.rst:33]: I thought this was already covered (both linux building as well as finding u-boot and kernel sources)

File: source/reference-manual/linux/linux-update.rst [hunk 23]
```diff
@@ -12,30 +12,515 @@ At Foundries.io™, we often `release`_ Linux® microPlatform (LmP) updates as a
...
+``meta-subscriber-overrides``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The previous step pushes the update to the FoundriesFactory, which triggers a new platform build for the latest LmP release.
+It is expected that, during the migration, this initial build often fails as ``meta-subcriber-overrides`` may have changes that require adjustments to the new release.
+
+.. tip::
+    It is helpful to :ref:`sync the Factory sources locally <ref-linux-building-ref>` so you can easily navigate through the Factory code.
+    Some of the suggestions here are based on this.
+    Remember: ``bitbake -e`` is a powerful tool.
```
Comment (vanmaegima) [inline source/reference-manual/linux/linux-update.rst:100]: @caiotpereira please check this note

**Pull request merged**